pkg/machines/components/storagePools/storagePoolDelete.jsx

```diff
@@ -0,0 +1,140 @@
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Button, Modal } from 'patternfly-react';
+
+import { storagePoolId } from '../../helpers.js';
+import { ModalError } from '../notification/inlineNotification.jsx';
+import { storagePoolDeactivate, storagePoolUndefine, storageVolumeDelete } from '../../libvirt-dbus.js';
+import cockpit from 'cockpit';
+
+const _ = cockpit.gettext;
+
+export class StoragePoolDelete extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            showModal: false,
+            dialogError: undefined,
+            deleteVolumes: false,
+        };
+        this.open = this.open.bind(this);
+        this.close = this.close.bind(this);
+        this.delete = this.delete.bind(this);
+        this.onValueChanged = this.onValueChanged.bind(this);
+        this.dialogErrorSet = this.dialogErrorSet.bind(this);
+    }
+
+    onValueChanged(key, value) {
+        const stateDelta = { [key]: value };
+
+        this.setState(stateDelta);
+    }
+
+    dialogErrorSet(text, detail) {
+        this.setState({ dialogError: text, dialogErrorDetail: detail });
+    }
+
+    close() {
+        this.setState({ showModal: false, dialogError: undefined });
+    }
+
+    open() {
+        this.setState({ showModal: true });
+    }
+
+    delete() {
+        const storagePool = this.props.storagePool;
+        const volumes = storagePool.volumes || [];
+        const storagePoolDeactivateAndUndefine = (storagePool) => {
+            if (storagePool.active) {
+                return storagePoolDeactivate(storagePool.connectionName, storagePool.id)
+                        .then(() => storagePoolUndefine(storagePool.connectionName, storagePool.id));
+            } else {
+                return storagePoolUndefine(storagePool.connectionName, storagePool.id);
+            }
+        };
+
+        if (this.state.deleteVolumes && storagePool.volumes.length > 0) {
+            Promise.all(volumes.map(volume => storageVolumeDelete(storagePool.connectionName, volume.path)))
+                    .then(() => storagePoolDeactivateAndUndefine(storagePool))
+                    .then(() => this.close,
+                          exc => this.dialogErrorSet(_("The Storage Pool could not be deleted"), exc.message));
+        } else {
+            storagePoolDeactivateAndUndefine(storagePool)
+                    .then(() => this.close,
+                          exc => this.dialogErrorSet(_("The Storage Pool could not be deleted"), exc.message));
+        }
+    }
+
+    render() {
+        const { storagePool } = this.props;
+        const id = storagePoolId(storagePool.name, storagePool.connectionName);
+
+        let defaultBody = (
+            <div className='ct-form-layout'>
+                { storagePool.active && storagePool.volumes && <React.Fragment>
```

Comment:
Should also test for `volumes.length > 0` here? (empty lists are truthy, irritatingly)

Reply:
Technically this will not cause problems because I don't add the volumes object at all if a pool doesn't have volumes. So if it exists it's always populated.
But I 'll replace it, it's not clear code indeed.